File: identity/src/test/java/com/android/identity/MobileSecurityObjectGeneratorTest.java
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity;

import android.security.keystore.KeyProperties;

import androidx.test.filters.SmallTest;

import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.spec.ECGenParameterSpec;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class MobileSecurityObjectGeneratorTest {
    private static final String MDL_DOCTYPE = "org.iso.18013.5.1.mDL";
    private static final String MDL_NAMESPACE = "org.iso.18013.5.1";
    private static final String AAMVA_NAMESPACE = "org.aamva.18013.5.1";

    private static final String MVR_DOCTYPE = "org.iso.18013.7.1.mVR";
    private static final String MVR_NAMESPACE = "org.iso.18013.7.1";

    private KeyPair generateReaderKeyPair() throws Exception {
        KeyPairGenerator kpg = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC);
        ECGenParameterSpec ecSpec = new ECGenParameterSpec("secp256r1");
        kpg.initialize(ecSpec);
        return kpg.generateKeyPair();
    }

    private Map<Long, byte[]> generateISODigest(String digestAlgorithm) throws NoSuchAlgorithmException {
        MessageDigest digester = MessageDigest.getInstance(digestAlgorithm);
        Map<Long, byte[]> isoDigestIDs = new HashMap<>();
        isoDigestIDs.put(0L, digester.digest("aardvark".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(1L, digester.digest("alligator".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(2L, digester.digest("baboon".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(3L, digester.digest("butterfly".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(4L, digester.digest("cat".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(5L, digester.digest("cricket".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(6L, digester.digest("dog".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(7L, digester.digest("elephant".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(8L, digester.digest("firefly".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(9L, digester.digest("frog".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(10L, digester.digest("gecko".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(11L, digester.digest("hippo".getBytes(StandardCharsets.UTF_8)));
        isoDigestIDs.put(12L, digester.digest("iguana".getBytes(StandardCharsets.UTF_8)));
        return isoDigestIDs;
    }

    private Map<Long, byte[]> generateISOUSDigest(String digestAlgorithm) throws NoSuchAlgorithmException {
        MessageDigest digester = MessageDigest.getInstance(digestAlgorithm);
        Map<Long, byte[]> isoUSDigestIDs = new HashMap<>();
        isoUSDigestIDs.put(0L, digester.digest("jaguar".getBytes(StandardCharsets.UTF_8)));
        isoUSDigestIDs.put(1L, digester.digest("jellyfish".getBytes(StandardCharsets.UTF_8)));
        isoUSDigestIDs.put(2L, digester.digest("koala".getBytes(StandardCharsets.UTF_8)));
        isoUSDigestIDs.put(3L, digester.digest("lemur".getBytes(StandardCharsets.UTF_8)));
        return isoUSDigestIDs;
    }

    private void checkISODigest(Map<Long, byte[]> isoDigestIDs, String digestAlgorithm)
            throws NoSuchAlgorithmException {
        MessageDigest digester = MessageDigest.getInstance(digestAlgorithm);
        Assert.assertEquals(Set.of(0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L, 11L, 12L),
                isoDigestIDs.keySet());
        Assert.assertArrayEquals(digester.digest("aardvark".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(0L));
        Assert.assertArrayEquals(digester.digest("alligator".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(1L));
        Assert.assertArrayEquals(digester.digest("baboon".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(2L));
        Assert.assertArrayEquals(digester.digest("butterfly".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(3L));
        Assert.assertArrayEquals(digester.digest("cat".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(4L));
        Assert.assertArrayEquals(digester.digest("cricket".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(5L));
        Assert.assertArrayEquals(digester.digest("dog".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(6L));
        Assert.assertArrayEquals(digester.digest("elephant".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(7L));
        Assert.assertArrayEquals(digester.digest("firefly".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(8L));
        Assert.assertArrayEquals(digester.digest("frog".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(9L));
        Assert.assertArrayEquals(digester.digest("gecko".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(10L));
        Assert.assertArrayEquals(digester.digest("hippo".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(11L));
        Assert.assertArrayEquals(digester.digest("iguana".getBytes(StandardCharsets.UTF_8)),
                isoDigestIDs.get(12L));
    }

    private void checkISOUSDigest(Map<Long, byte[]> isoUSDigestIDs, String digestAlgorithm)
            throws NoSuchAlgorithmException {
        MessageDigest digester = MessageDigest.getInstance(digestAlgorithm);
        Assert.assertEquals(Set.of(0L, 1L, 2L, 3L), isoUSDigestIDs.keySet());
        Assert.assertArrayEquals(digester.digest("jaguar".getBytes(StandardCharsets.UTF_8)),
                isoUSDigestIDs.get(0L));
        Assert.assertArrayEquals(digester.digest("jellyfish".getBytes(StandardCharsets.UTF_8)),
                isoUSDigestIDs.get(1L));
        Assert.assertArrayEquals(digester.digest("koala".getBytes(StandardCharsets.UTF_8)),
                isoUSDigestIDs.get(2L));
        Assert.assertArrayEquals(digester.digest("lemur".getBytes(StandardCharsets.UTF_8)),
                isoUSDigestIDs.get(3L));
    }

    public void testFullMSO(String digestAlgorithm) throws NoSuchAlgorithmException {
        PublicKey deviceKeyFromVector = Util.getPublicKeyFromIntegers(
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X, 16),
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y, 16));
        final Timestamp signedTimestamp = Timestamp.ofEpochMilli(1601559002000L);
        final Timestamp validFromTimestamp = Timestamp.ofEpochMilli(1601559002000L);
        final Timestamp validUntilTimestamp = Timestamp.ofEpochMilli(1633095002000L);
        final Timestamp expectedTimestamp = Timestamp.ofEpochMilli(1611093002000L);

        Map<String, List<String>> deviceKeyAuthorizedDataElements = new HashMap<>();
        deviceKeyAuthorizedDataElements.put("a", List.of("1", "2", "f"));
        deviceKeyAuthorizedDataElements.put("b", List.of("4", "5", "k"));

        Map<Long, byte[]> keyInfo = new HashMap<>();
        keyInfo.put(10L, Util.fromHex("C985"));

        byte[] encodedMSO = new MobileSecurityObjectGenerator(digestAlgorithm,
                "org.iso.18013.5.1.mDL", deviceKeyFromVector)
                .addDigestIdsForNamespace("org.iso.18013.5.1", generateISODigest(digestAlgorithm))
                .addDigestIdsForNamespace("org.iso.18013.5.1.US", generateISOUSDigest(digestAlgorithm))
                .setDeviceKeyAuthorizedNameSpaces(List.of("abc", "bcd"))
                .setDeviceKeyAuthorizedDataElements(deviceKeyAuthorizedDataElements)
                .setDeviceKeyInfo(keyInfo)
                .setValidityInfo(
                        signedTimestamp,
                        validFromTimestamp,
                        validUntilTimestamp,
                        expectedTimestamp)
                .generate();

        MobileSecurityObjectParser.MobileSecurityObject mso = new MobileSecurityObjectParser()
                .setMobileSecurityObject(encodedMSO).parse();

        Assert.assertEquals("1.0", mso.getVersion());
        Assert.assertEquals(digestAlgorithm, mso.getDigestAlgorithm());
        Assert.assertEquals("org.iso.18013.5.1.mDL", mso.getDocType());

        Assert.assertEquals(Set.of("org.iso.18013.5.1", "org.iso.18013.5.1.US"),
                mso.getValueDigestNamespaces());
        Assert.assertNull(mso.getDigestIDs("abc"));
        checkISODigest(mso.getDigestIDs("org.iso.18013.5.1"), digestAlgorithm);
        checkISOUSDigest(mso.getDigestIDs("org.iso.18013.5.1.US"), digestAlgorithm);

        Assert.assertEquals(deviceKeyFromVector, mso.getDeviceKey());
        Assert.assertEquals(List.of("abc", "bcd"), mso.getDeviceKeyAuthorizedNameSpaces());
        Assert.assertEquals(deviceKeyAuthorizedDataElements, mso.getDeviceKeyAuthorizedDataElements());
        Assert.assertEquals(keyInfo.keySet(), mso.getDeviceKeyInfo().keySet());
        Assert.assertEquals(Util.toHex(keyInfo.get(10L)), Util.toHex(mso.getDeviceKeyInfo().get(10L)));

        Assert.assertEquals(signedTimestamp, mso.getSigned());
        Assert.assertEquals(validFromTimestamp, mso.getValidFrom());
        Assert.assertEquals(validUntilTimestamp, mso.getValidUntil());
        Assert.assertEquals(expectedTimestamp, mso.getExpectedUpdate());
    }

    @Test
    @SmallTest
    public void testBasicMSO() throws Exception {
        PublicKey deviceKeyFromVector = Util.getPublicKeyFromIntegers(
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X, 16),
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y, 16));
        final Timestamp signedTimestamp = Timestamp.ofEpochMilli(1601559002000L);
        final Timestamp validFromTimestamp = Timestamp.ofEpochMilli(1601559002000L);
        final Timestamp validUntilTimestamp = Timestamp.ofEpochMilli(1633095002000L);
        final String digestAlgorithm = "SHA-256";

        byte[] encodedMSO = new MobileSecurityObjectGenerator(digestAlgorithm,
                "org.iso.18013.5.1.mDL", deviceKeyFromVector)
                .addDigestIdsForNamespace("org.iso.18013.5.1", generateISODigest(digestAlgorithm))
                .addDigestIdsForNamespace("org.iso.18013.5.1.US", generateISOUSDigest(digestAlgorithm))
                .setValidityInfo(signedTimestamp, validFromTimestamp, validUntilTimestamp, null)
                .generate();

        MobileSecurityObjectParser.MobileSecurityObject mso = new MobileSecurityObjectParser()
                .setMobileSecurityObject(encodedMSO).parse();

        Assert.assertEquals("1.0", mso.getVersion());
        Assert.assertEquals(digestAlgorithm, mso.getDigestAlgorithm());
        Assert.assertEquals("org.iso.18013.5.1.mDL", mso.getDocType());

        Assert.assertEquals(Set.of("org.iso.18013.5.1", "org.iso.18013.5.1.US"),
                mso.getValueDigestNamespaces());
        Assert.assertNull(mso.getDigestIDs("abc"));
        checkISODigest(mso.getDigestIDs("org.iso.18013.5.1"), digestAlgorithm);
        checkISOUSDigest(mso.getDigestIDs("org.iso.18013.5.1.US"), digestAlgorithm);

        Assert.assertEquals(deviceKeyFromVector, mso.getDeviceKey());
        Assert.assertNull(mso.getDeviceKeyAuthorizedNameSpaces());
        Assert.assertNull(mso.getDeviceKeyAuthorizedDataElements());
        Assert.assertNull(mso.getDeviceKeyInfo());

        Assert.assertEquals(signedTimestamp, mso.getSigned());
        Assert.assertEquals(validFromTimestamp, mso.getValidFrom());
        Assert.assertEquals(validUntilTimestamp, mso.getValidUntil());
        Assert.assertNull(mso.getExpectedUpdate());
    }

    @Test
    @SmallTest
    public void testFullMSO_Sha256() throws NoSuchAlgorithmException {
        testFullMSO("SHA-256");
    }

    @Test
    @SmallTest
    public void testFullMSO_Sha384() throws NoSuchAlgorithmException {
        testFullMSO("SHA-384");
    }

    @Test
    @SmallTest
    public void testFullMSO_Sha512() throws NoSuchAlgorithmException {
        testFullMSO("SHA-512");
    }

    @Test
    @SmallTest
    public void testMSOExceptions() {
        PublicKey deviceKeyFromVector = Util.getPublicKeyFromIntegers(
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X, 16),
                new BigInteger(TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y, 16));

        Assert.assertThrows("expect exception for illegal digestAlgorithm",
                IllegalArgumentException.class,
                () -> new MobileSecurityObjectGenerator("SHA-257",
                        "org.iso.18013.5.1.mDL", deviceKeyFromVector));

        final String digestAlgorithm = "SHA-256";
        MobileSecurityObjectGenerator msoGenerator = new MobileSecurityObjectGenerator(digestAlgorithm,
                "org.iso.18013.5.1.mDL", deviceKeyFromVector);

        Assert.assertThrows("expect exception for empty digestIDs",
                IllegalArgumentException.class,
                () -> msoGenerator.addDigestIdsForNamespace("org.iso.18013.5.1", new HashMap<>()));

        Assert.assertThrows("expect exception for validFrom < signed",
                IllegalArgumentException.class,
                () -> msoGenerator.setValidityInfo(
                        Timestamp.ofEpochMilli(1601559002000L),
                        Timestamp.ofEpochMilli(1601559001000L),
                        Timestamp.ofEpochMilli(1633095002000L),
                        Timestamp.ofEpochMilli(1611093002000L)));

        Assert.assertThrows("expect exception for validUntil <= validFrom",
                IllegalArgumentException.class,
                () -> msoGenerator.setValidityInfo(
                        Timestamp.ofEpochMilli(1601559002000L),
                        Timestamp.ofEpochMilli(1601559002000L),
                        Timestamp.ofEpochMilli(1601559002000L),
                        Timestamp.ofEpochMilli(1611093002000L)));

        Map<String, List<String>> deviceKeyAuthorizedDataElements = new HashMap<>();
        deviceKeyAuthorizedDataElements.put("a", List.of("1", "2", "f"));
        deviceKeyAuthorizedDataElements.put("b", List.of("4", "5", "k"));
        Assert.assertThrows("expect exception for deviceKeyAuthorizedDataElements including " +
                        "namespace in deviceKeyAuthorizedNameSpaces",
                IllegalArgumentException.class,
                () -> msoGenerator.setDeviceKeyAuthorizedNameSpaces(List.of("a", "bcd"))
                        .setDeviceKeyAuthorizedDataElements(deviceKeyAuthorizedDataElements));
        Assert.assertThrows("expect exception for deviceKeyAuthorizedNameSpaces including " +
                        "namespace in deviceKeyAuthorizedDataElements",
                IllegalArgumentException.class,
                () -> msoGenerator.setDeviceKeyAuthorizedDataElements(deviceKeyAuthorizedDataElements)
                        .setDeviceKeyAuthorizedNameSpaces(List.of("a", "bcd")));

        Assert.assertThrows("expect exception for msoGenerator which has not had " +
                        "addDigestIdsForNamespace and setValidityInfo called before generating",
                IllegalStateException.class,
                msoGenerator::generate);


        Assert.assertThrows("expect exception for msoGenerator which has not had " +
                        "addDigestIdsForNamespace called before generating",
                IllegalStateException.class,
                () -> {new MobileSecurityObjectGenerator(digestAlgorithm,
                        "org.iso.18013.5.1.mDL", deviceKeyFromVector)
                        .setValidityInfo(
                                Timestamp.ofEpochMilli(1601559002000L),
                                Timestamp.ofEpochMilli(1601559002000L),
                                Timestamp.ofEpochMilli(1633095002000L),
                                Timestamp.ofEpochMilli(1611093002000L))
                        .generate();});

        Assert.assertThrows("expect exception for msoGenerator which has not had " +
                        "setValidityInfo called before generating",
                IllegalStateException.class,
                () -> {new MobileSecurityObjectGenerator(digestAlgorithm,
                        "org.iso.18013.5.1.mDL", deviceKeyFromVector)
                        .addDigestIdsForNamespace("org.iso.18013.5.1", generateISODigest(digestAlgorithm))
                        .addDigestIdsForNamespace("org.iso.18013.5.1.US", generateISOUSDigest(digestAlgorithm))
                        .generate();});

    }
}

File: identity-mdoc/src/test/java/com/android/identity/mdoc/mso/MobileSecurityObjectGeneratorTest.kt
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.mdoc.mso

import com.android.identity.crypto.Algorithm
import com.android.identity.crypto.Crypto
import com.android.identity.crypto.EcCurve
import com.android.identity.crypto.EcPublicKeyDoubleCoordinate
import com.android.identity.mdoc.TestVectors
import com.android.identity.util.fromHex
import com.android.identity.util.toHex
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.security.Security
import kotlinx.datetime.Instant

class MobileSecurityObjectGeneratorTest {
    @Before
    fun setup() {
        Security.insertProviderAt(BouncyCastleProvider(), 1)
    }

    private fun getDigestAlg(digestAlgorithm: String): Algorithm {
        return when (digestAlgorithm) {
            "SHA-256" -> Algorithm.SHA256
            "SHA-384" -> Algorithm.SHA384
            "SHA-512" -> Algorithm.SHA512
            else -> throw AssertionError()
        }
    }
    
    private fun generateISODigest(digestAlgorithm: String): Map<Long, ByteArray> {
        val alg = getDigestAlg(digestAlgorithm)
        val isoDigestIDs = mutableMapOf<Long, ByteArray>()
        isoDigestIDs[0L] = Crypto.digest(alg, "aardvark".toByteArray())
        isoDigestIDs[1L] = Crypto.digest(alg, "alligator".toByteArray())
        isoDigestIDs[2L] = Crypto.digest(alg, "baboon".toByteArray())
        isoDigestIDs[3L] = Crypto.digest(alg, "butterfly".toByteArray())
        isoDigestIDs[4L] = Crypto.digest(alg, "cat".toByteArray())
        isoDigestIDs[5L] = Crypto.digest(alg, "cricket".toByteArray())
        isoDigestIDs[6L] = Crypto.digest(alg, "dog".toByteArray())
        isoDigestIDs[7L] = Crypto.digest(alg, "elephant".toByteArray())
        isoDigestIDs[8L] = Crypto.digest(alg, "firefly".toByteArray())
        isoDigestIDs[9L] = Crypto.digest(alg, "frog".toByteArray())
        isoDigestIDs[10L] = Crypto.digest(alg, "gecko".toByteArray())
        isoDigestIDs[11L] = Crypto.digest(alg, "hippo".toByteArray())
        isoDigestIDs[12L] = Crypto.digest(alg, "iguana".toByteArray())
        return isoDigestIDs
    }

    private fun generateISOUSDigest(digestAlgorithm: String): Map<Long, ByteArray> {
        val alg = getDigestAlg(digestAlgorithm)
        val isoUSDigestIDs: MutableMap<Long, ByteArray> = HashMap()
        isoUSDigestIDs[0L] = Crypto.digest(alg, "jaguar".toByteArray())
        isoUSDigestIDs[1L] = Crypto.digest(alg, "jellyfish".toByteArray())
        isoUSDigestIDs[2L] = Crypto.digest(alg, "koala".toByteArray())
        isoUSDigestIDs[3L] = Crypto.digest(alg, "lemur".toByteArray())
        return isoUSDigestIDs
    }

    private fun checkISODigest(isoDigestIDs: Map<Long, ByteArray>?, digestAlgorithm: String) {
        val alg = getDigestAlg(digestAlgorithm)
        Assert.assertEquals(
            setOf(0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L, 11L, 12L),
            isoDigestIDs!!.keys
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "aardvark".toByteArray()),
            isoDigestIDs[0L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "alligator".toByteArray()),
            isoDigestIDs[1L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "baboon".toByteArray()),
            isoDigestIDs[2L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "butterfly".toByteArray()),
            isoDigestIDs[3L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "cat".toByteArray()),
            isoDigestIDs[4L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "cricket".toByteArray()),
            isoDigestIDs[5L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "dog".toByteArray()),
            isoDigestIDs[6L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "elephant".toByteArray()),
            isoDigestIDs[7L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "firefly".toByteArray()),
            isoDigestIDs[8L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "frog".toByteArray()),
            isoDigestIDs[9L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "gecko".toByteArray()),
            isoDigestIDs[10L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "hippo".toByteArray()),
            isoDigestIDs[11L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "iguana".toByteArray()),
            isoDigestIDs[12L]
        )
    }

    private fun checkISOUSDigest(isoUSDigestIDs: Map<Long, ByteArray>?, digestAlgorithm: String) {
        val alg = getDigestAlg(digestAlgorithm)
        Assert.assertEquals(setOf(0L, 1L, 2L, 3L), isoUSDigestIDs!!.keys)
        Assert.assertArrayEquals(
            Crypto.digest(alg, "jaguar".toByteArray()),
            isoUSDigestIDs[0L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "jellyfish".toByteArray()),
            isoUSDigestIDs[1L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "koala".toByteArray()),
            isoUSDigestIDs[2L]
        )
        Assert.assertArrayEquals(
            Crypto.digest(alg, "lemur".toByteArray()),
            isoUSDigestIDs[3L]
        )
    }

    fun testFullMSO(digestAlgorithm: String) {
        val deviceKeyFromVector = EcPublicKeyDoubleCoordinate(
            EcCurve.P256,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X.fromHex,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y.fromHex
        )
        val signedTimestamp = Instant.fromEpochMilliseconds(1601559002000L)
        val validFromTimestamp = Instant.fromEpochMilliseconds(1601559002000L)
        val validUntilTimestamp = Instant.fromEpochMilliseconds(1633095002000L)
        val expectedTimestamp = Instant.fromEpochMilliseconds(1611093002000L)
        val deviceKeyAuthorizedDataElements: MutableMap<String, List<String>> = HashMap()
        deviceKeyAuthorizedDataElements["a"] = listOf("1", "2", "f")
        deviceKeyAuthorizedDataElements["b"] = listOf("4", "5", "k")
        val keyInfo: MutableMap<Long, ByteArray> = HashMap()
        keyInfo[10L] = "C985".fromHex
        val encodedMSO = MobileSecurityObjectGenerator(
            digestAlgorithm,
            "org.iso.18013.5.1.mDL", deviceKeyFromVector
        )
            .addDigestIdsForNamespace("org.iso.18013.5.1", generateISODigest(digestAlgorithm))
            .addDigestIdsForNamespace("org.iso.18013.5.1.US", generateISOUSDigest(digestAlgorithm))
            .setDeviceKeyAuthorizedNameSpaces(listOf("abc", "bcd"))
            .setDeviceKeyAuthorizedDataElements(deviceKeyAuthorizedDataElements)
            .setDeviceKeyInfo(keyInfo)
            .setValidityInfo(
                signedTimestamp,
                validFromTimestamp,
                validUntilTimestamp,
                expectedTimestamp
            )
            .generate()
        val mso = MobileSecurityObjectParser(
            encodedMSO
        ).parse()
        Assert.assertEquals("1.0", mso.version)
        Assert.assertEquals(digestAlgorithm, mso.digestAlgorithm)
        Assert.assertEquals("org.iso.18013.5.1.mDL", mso.docType)
        Assert.assertEquals(
            setOf("org.iso.18013.5.1", "org.iso.18013.5.1.US"),
            mso.valueDigestNamespaces
        )
        Assert.assertNull(mso.getDigestIDs("abc"))
        checkISODigest(mso.getDigestIDs("org.iso.18013.5.1"), digestAlgorithm)
        checkISOUSDigest(mso.getDigestIDs("org.iso.18013.5.1.US"), digestAlgorithm)
        Assert.assertEquals(deviceKeyFromVector, mso.deviceKey)
        Assert.assertEquals(listOf("abc", "bcd"), mso.deviceKeyAuthorizedNameSpaces)
        Assert.assertEquals(deviceKeyAuthorizedDataElements, mso.deviceKeyAuthorizedDataElements)
        Assert.assertEquals(keyInfo.keys, mso.deviceKeyInfo!!.keys)
        Assert.assertEquals(
            keyInfo[10L]!!.toHex,
            mso.deviceKeyInfo!![10L]!!.toHex
        )
        Assert.assertEquals(signedTimestamp, mso.signed)
        Assert.assertEquals(validFromTimestamp, mso.validFrom)
        Assert.assertEquals(validUntilTimestamp, mso.validUntil)
        Assert.assertEquals(expectedTimestamp, mso.expectedUpdate)
    }

    @Test
    fun testBasicMSO() {
        val deviceKeyFromVector = EcPublicKeyDoubleCoordinate(
            EcCurve.P256,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X.fromHex,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y.fromHex
        )
        val signedTimestamp = Instant.fromEpochMilliseconds(1601559002000L)
        val validFromTimestamp = Instant.fromEpochMilliseconds(1601559002000L)
        val validUntilTimestamp = Instant.fromEpochMilliseconds(1633095002000L)
        val digestAlgorithm = "SHA-256"
        val encodedMSO = MobileSecurityObjectGenerator(
            digestAlgorithm,
            "org.iso.18013.5.1.mDL", deviceKeyFromVector
        )
            .addDigestIdsForNamespace("org.iso.18013.5.1", generateISODigest(digestAlgorithm))
            .addDigestIdsForNamespace("org.iso.18013.5.1.US", generateISOUSDigest(digestAlgorithm))
            .setValidityInfo(signedTimestamp, validFromTimestamp, validUntilTimestamp, null)
            .generate()
        val mso = MobileSecurityObjectParser(encodedMSO).parse()
        Assert.assertEquals("1.0", mso.version)
        Assert.assertEquals(digestAlgorithm, mso.digestAlgorithm)
        Assert.assertEquals("org.iso.18013.5.1.mDL", mso.docType)
        Assert.assertEquals(
            setOf("org.iso.18013.5.1", "org.iso.18013.5.1.US"),
            mso.valueDigestNamespaces
        )
        Assert.assertNull(mso.getDigestIDs("abc"))
        checkISODigest(mso.getDigestIDs("org.iso.18013.5.1"), digestAlgorithm)
        checkISOUSDigest(mso.getDigestIDs("org.iso.18013.5.1.US"), digestAlgorithm)
        Assert.assertEquals(deviceKeyFromVector, mso.deviceKey)
        Assert.assertNull(mso.deviceKeyAuthorizedNameSpaces)
        Assert.assertNull(mso.deviceKeyAuthorizedDataElements)
        Assert.assertNull(mso.deviceKeyInfo)
        Assert.assertEquals(signedTimestamp, mso.signed)
        Assert.assertEquals(validFromTimestamp, mso.validFrom)
        Assert.assertEquals(validUntilTimestamp, mso.validUntil)
        Assert.assertNull(mso.expectedUpdate)
    }

    @Test
    fun testFullMSO_Sha256() {
        testFullMSO("SHA-256")
    }

    @Test
    fun testFullMSO_Sha384() {
        testFullMSO("SHA-384")
    }

    @Test
    fun testFullMSO_Sha512() {
        testFullMSO("SHA-512")
    }

    @Test
    fun testMSOExceptions() {
        val deviceKeyFromVector = EcPublicKeyDoubleCoordinate(
            EcCurve.P256,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_X.fromHex,
            TestVectors.ISO_18013_5_ANNEX_D_STATIC_DEVICE_KEY_Y.fromHex
        )
        Assert.assertThrows(
            "expect exception for illegal digestAlgorithm",
            IllegalArgumentException::class.java
        ) {
            MobileSecurityObjectGenerator(
                "SHA-257",
                "org.iso.18013.5.1.mDL", deviceKeyFromVector
            )
        }
        val digestAlgorithm = "SHA-256"
        val msoGenerator = MobileSecurityObjectGenerator(
            digestAlgorithm,
            "org.iso.18013.5.1.mDL", deviceKeyFromVector
        )
        Assert.assertThrows(
            "expect exception for empty digestIDs",
            IllegalArgumentException::class.java
        ) { msoGenerator.addDigestIdsForNamespace("org.iso.18013.5.1", HashMap()) }
        Assert.assertThrows(
            "expect exception for validFrom < signed",
            IllegalArgumentException::class.java
        ) {
            msoGenerator.setValidityInfo(
                Instant.fromEpochMilliseconds(1601559002000L),
                Instant.fromEpochMilliseconds(1601559001000L),
                Instant.fromEpochMilliseconds(1633095002000L),
                Instant.fromEpochMilliseconds(1611093002000L)
            )
        }
        Assert.assertThrows(
            "expect exception for validUntil <= validFrom",
            IllegalArgumentException::class.java
        ) {
            msoGenerator.setValidityInfo(
                Instant.fromEpochMilliseconds(1601559002000L),
                Instant.fromEpochMilliseconds(1601559002000L),
                Instant.fromEpochMilliseconds(1601559002000L),
                Instant.fromEpochMilliseconds(1611093002000L)
            )
        }
        val deviceKeyAuthorizedDataElements: MutableMap<String, List<String>> = HashMap()
        deviceKeyAuthorizedDataElements["a"] = listOf("1", "2", "f")
        deviceKeyAuthorizedDataElements["b"] = listOf("4", "5", "k")
        Assert.assertThrows(
            "expect exception for deviceKeyAuthorizedDataElements including " +
                    "namespace in deviceKeyAuthorizedNameSpaces",
            IllegalArgumentException::class.java
        ) {
            msoGenerator.setDeviceKeyAuthorizedNameSpaces(listOf("a", "bcd"))
                .setDeviceKeyAuthorizedDataElements(deviceKeyAuthorizedDataElements)
        }
        Assert.assertThrows(
            "expect exception for deviceKeyAuthorizedNameSpaces including " +
                    "namespace in deviceKeyAuthorizedDataElements",
            IllegalArgumentException::class.java
        ) {
            msoGenerator.setDeviceKeyAuthorizedDataElements(deviceKeyAuthorizedDataElements)
                .setDeviceKeyAuthorizedNameSpaces(listOf("a", "bcd"))
        }
        Assert.assertThrows(
            "expect exception for msoGenerator which has not had " +
                    "addDigestIdsForNamespace and setValidityInfo called before generating",
            IllegalStateException::class.java
        ) { msoGenerator.generate() }
        Assert.assertThrows(
            "expect exception for msoGenerator which has not had " +
                    "addDigestIdsForNamespace called before generating",
            IllegalStateException::class.java
        ) {
            MobileSecurityObjectGenerator(
                digestAlgorithm,
                "org.iso.18013.5.1.mDL", deviceKeyFromVector
            )
                .setValidityInfo(
                    Instant.fromEpochMilliseconds(1601559002000L),
                    Instant.fromEpochMilliseconds(1601559002000L),
                    Instant.fromEpochMilliseconds(1633095002000L),
                    Instant.fromEpochMilliseconds(1611093002000L)
                )
                .generate()
        }
        Assert.assertThrows(
            "expect exception for msoGenerator which has not had " +
                    "setValidityInfo called before generating",
            IllegalStateException::class.java
        ) {
            MobileSecurityObjectGenerator(
                digestAlgorithm,
                "org.iso.18013.5.1.mDL", deviceKeyFromVector
            )
                .addDigestIdsForNamespace("org.iso.18013.5.1", generateISODigest(digestAlgorithm))
                .addDigestIdsForNamespace(
                    "org.iso.18013.5.1.US",
                    generateISOUSDigest(digestAlgorithm)
                )
                .generate()
        }
    }
}
